File: leetcode/C/2.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
 #include "stdlib.h"
 struct ListNode {
  	int val;
    struct ListNode *next;
 };
struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {
    ListNode *head1 = l1;
    ListNode *head2 = l2;
    ListNode *list = (ListNode *)malloc(sizeof(struct ListNode));
    list->val = 0;
    list->next = NULL;
    ListNode *tmp = list;
    while(head1 != NULL && head2 != NULL) {
    	ListNode *node = (ListNode*)malloc(sizeof(struct ListNode));
    	node->val = 0;
    	node->next = NULL;
    	if (tmp->next)
    	{
    		node->val = tmp->next->val;
    	}
    	int sum =  head1->val + head2->val+node->val;
    	if (sum>=10)
    	 {
    	 	ListNode *node1 = (ListNode*)malloc(sizeof(struct ListNode));
    		node1->val = sum/10;
    		node1->next = NULL;
    		node->next = node1;
    	 } 
    	node->val = sum%10;
    	tmp->next = node;
    	head1 = head1->next;
    	head2 = head2->next;
    }
    return list->next;
}

File: leetcode/Go/2.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

type ListNode struct {
   Val int
   Next *ListNode
}
// func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
// 	l3 := l1
//     for {
// 		if l1 == nil {
// 			return l3
// 		}
// 		if l2 != nil {
// 			l1.Val = l1.Val+l2.Val
// 			l2 = l2.Next
// 		}
// 		if l1.Val > 9 {
// 			val := l1.Val
// 			l1.Val = val%10
// 			if l1.Next == nil {
// 				l1.Next = new(ListNode)
// 			}
// 			l1.Next.Val += val/10
// 		}
// 		if l1.Next == nil && l2 != nil {
// 			l1.Next = l2
// 			return l3
// 		}
// 		l1 = l1.Next
// 	}
// }

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	head := new(ListNode)
	current := head
	tmp := 0
    for l1 != nil || l2 != nil || tmp >0 {
		sum := tmp
		if l1 != nil {
			sum += l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			sum += l2.Val
			l2 = l2.Next
		}

		tmp = sum / 10
		current.Val = sum %10
		if l1 ==nil && l2 == nil && tmp == 0 {
			return head
		}
		if current.Next == nil {
			current.Next = new(ListNode)
		}
		current = current.Next
		if tmp > 0{
			current.Val = tmp
		}
	}
	return head
}